<a href="https://colab.research.google.com/github/lwy2107/GK_co_work/blob/master/%EC%8B%A0%EA%B2%BD%EB%A7%9D_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [9]:
from tensorflow import keras
(train_input, train_target), (test_input, test_target)=\
keras.datasets.fashion_mnist.load_data()

4422102/4422102 [==============================] - 0s 0us/step


In [10]:
from sklearn.model_selection import train_test_split
train_scaled=train_input/255.0
train_scaled=train_scaled.reshape(-1,28*28)
train_scaled, val_scaled, train_target, val_target=train_test_split(
    train_scaled, train_target, test_size=0.2, random_state=42)

In [11]:
dense1=keras.layers.Dense(100, activation='sigmoid', input_shape=(784,))#시그모이드 함수를 사용 신경망은 첫번째층에 무조건 input_shape매개변수크기를 입력해야함
dense2=keras.layers.Dense(10, activation='softmax') # 소프트맥스 층 선언

In [12]:
model = keras.Sequential([dense1, dense2]) #심층 신경망 사용 첫째층은 시그모이드 두번째층은 소프트맥스층을 사용하였다.

In [13]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 100)               78500     
                                                                 
 dense_1 (Dense)             (None, 10)                1010      
                                                                 
Total params: 79510 (310.59 KB)
Trainable params: 79510 (310.59 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


지금 summury메서드를 이용하여 출력을 해보았다.
모델은 sequential이라는 이름을 사용하고 있고 층은 총 2개의 층으로 구성된 딥러닝형식으로 구성되어있고 출력형식은 첫번째 층에서는 데이터셋의 데이터를를 입력하여 100개의 값을 출력하도록 하고 그 100개의 출력된값을 다시 두번째층에서 계산하여 10개의 결과값을 내도록하였다. 이를 파라미터로 나타냈다. 첫번째층에서는 784픽셀의 데이터에 100개의 뉴런 100개의 절편을 계산하게되는데
784*100+100 으로구성되어있다. 즉 78500개라는 파라미터 값이 나오게된다.
그다음층에서는 100개의 계산된 결과값을 받았고 10개의 뉴런의 값을 출력하게 하고 그에 맞는 절편을 구성하게된다.
100*10+10로 구성된다. 그 결과 파라미터값은 1010으로 나오게 되고 모든값을 더하게되면 79510개의 파라미터 값이 나오게된다.

In [15]:
model = keras.Sequential([
    keras.layers.Dense(100,activation='sigmoid', input_shape=(784,),name='hidden'),
    keras.layers.Dense(10, activation='softmax', name='output')], name='패션MNIST 모델')
model.summary()

Model: "패션MNIST 모델"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 hidden (Dense)              (None, 100)               78500     
                                                                 
 output (Dense)              (None, 10)                1010      
                                                                 
Total params: 79510 (310.59 KB)
Trainable params: 79510 (310.59 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [22]:
model = keras.Sequential()
model.add(keras.layers.Dense(100, activation='sigmoid', input_shape=(784,)))
model.add(keras.layers.Dense(10, activation='softmax'))
model.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_4 (Dense)             (None, 100)               78500     
                                                                 
 dense_5 (Dense)             (None, 10)                1010      
                                                                 
Total params: 79510 (310.59 KB)
Trainable params: 79510 (310.59 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [23]:
model.compile(loss='sparse_categorical_crossentropy', metrics='accuracy')
model.fit(train_scaled, train_target, epochs=5)

Epoch 1/5
1500/1500 [==============================] - 4s 2ms/step - loss: 0.5667 - accuracy: 0.8066
Epoch 2/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.4116 - accuracy: 0.8511
Epoch 3/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.3769 - accuracy: 0.8630
Epoch 4/5
1500/1500 [==============================] - 5s 3ms/step - loss: 0.3546 - accuracy: 0.8704
Epoch 5/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.3366 - accuracy: 0.8776


위의 신경망을 에포크 5회로 훈련시켜보았다.
에포크가 가면갈수록 예측의 정확도가 올라가는것을 알수 있다.

현재 사용중인 선형함수들은 단점이 있다. 0~1이라는 소수점을 모두 계산해야하다보니 정확한 값을 내기가 힘들어진다. 그러므로 렐루함수를 사용해서 0or1로 하여 결과를 나타내게 하면 보다 효과적으로 결과값을 도출해낼수 있다. 그리고 사용량도 줄어들게 된다.

AttributeError: 'Sequential' object has no attribute 'Sequential'